src/plugins/kibana_utils/public/state_management/url/url_storage.ts

```diff
@@ -0,0 +1,214 @@
+  const updateQueue: Array<(currentUrl: string) => string> = [];
+
+  // if we should replace or push with next async update,
+  // if any call in a queue asked to push, then we should push
+  let shouldReplace = true;
+
+  return {
+    listen: (cb: () => void) =>
+      history.listen(() => {
+        cb();
+      }),
+    update: (newUrl: string, replace = false) => updateUrl(newUrl, replace),
+    updateAsync: (updater: (currentUrl: string) => string, replace = false) => {
+      updateQueue.push(updater);
+      if (shouldReplace) {
+        shouldReplace = replace;
+      }
+
+      // Schedule url update to the next microtask
+      return Promise.resolve().then(() => {
+        if (updateQueue.length === 0) return getCurrentUrl();
+        const resultUrl = updateQueue.reduce((url, nextUpdate) => nextUpdate(url), getCurrentUrl());
+        const newUrl = updateUrl(resultUrl, shouldReplace);
+        // queue clean up
+        updateQueue.splice(0, updateQueue.length);
+        shouldReplace = true;
+
+        return newUrl;
+      });
+    },
+  };
+
+  function updateUrl(newUrl: string, replace = false): string {
+    const currentUrl = getCurrentUrl();
+    if (newUrl === currentUrl) return currentUrl; // skip update
+
+    const historyPath = getRelativeToHistoryPath(newUrl, history);
+
+    if (replace) {
+      history.replace(historyPath);
+    } else {
+      history.push(historyPath);
+    }
+
+    return getCurrentUrl();
+  }
+};
+
+/**
+ * Depending on history configuration extracts relative path for history updates
+ * 4 possible cases (see tests):
+ * 1. Browser history with empty base path
+ * 2. Browser history with base path
+ * 3. Hash history with empty base path
+ * 4. Hash history with base path
+ */
+export function getRelativeToHistoryPath(absoluteUrl: string, history: History): History.Path {
+  const parsedUrl = isHashHistory() ? parseUrlHash(absoluteUrl)! : parseUrl(absoluteUrl);
+  const parsedHash = isHashHistory() ? null : parseUrlHash(absoluteUrl);
+
+  return formatUrl({
+    pathname: stripBasename(parsedUrl.pathname),
+    search: stringifyQueryString(parsedUrl.query),
+    hash: parsedHash
+      ? formatUrl({
+          pathname: parsedHash.pathname,
+          search: stringifyQueryString(parsedHash.query),
+        })
+      : parsedUrl.hash,
+  });
+
+  function isHashHistory() {
+    return history.createHref({}).includes('#');
+  }
+
+  function stripBasename(path: string = '') {
+    const getBaseName = () => stripLeadingHash(stripTrailingSlash(history.createHref({})));
+    const baseName = getBaseName();
+    return path.startsWith(baseName) ? path.substr(baseName.length) : path;
+  }
+
+  function stripLeadingHash(path: string) {
+    return path.charAt(0) === '#' ? path.substr(1) : path;
+  }
+  function stripTrailingSlash(path: string) {
+    return path.charAt(path.length - 1) === '/' ? path.substr(0, path.length - 1) : path;
+  }
```

Comment:
I usually keep those underneath, just like I would keep private class members below public in class. But I agree that this could be confusing. Moved to the top.